How to attach an existing local resource to lambda?

Hi, 

I don't know how to attach an existing local resource to greengrass lambda through configuration of GGP.

When a local resource to be shared in several lambdas is defined in function.defaults.conf like this, error regarding duplication occurs.

$ ./ggp.sh -a X86_64 -g ggptest -d deployments/origins.conf

- function.defaults.conf
```
conf {
  ...
  localDeviceResources = []
  localVolumeResources = [
    {
      name="test-storage"
      sourcePath = "/test"
      destinationPath = "/test" 
      readWrite = true
    }
  ]
  ... 
```

- origins.conf
```
conf {
  functions = [
    "testFunction1",
    "testFunction2"
  ]
}
```

- console log
```
[INFO] BasicDeploymentHelper: Creating resource definition
[ERROR] BasicGreengrassHelper: Duplicate local volume resource source paths defined ["/test"].  Greengrass will not accept this configuration.
[ERROR] AwsGreengrassProvisioner: Invalid resource configuration
java.lang.RuntimeException: Invalid resource configuration
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicGreengrassHelper.disallowDuplicateLocalVolumeSourcePaths(BasicGreengrassHelper.java:940)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicGreengrassHelper.validateResourceDefinitionVersion(BasicGreengrassHelper.java:869)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicGreengrassHelper.createResourceDefinitionVersion(BasicGreengrassHelper.java:710)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicDeploymentHelper.execute(BasicDeploymentHelper.java:689)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicDeploymentHelper.execute(BasicDeploymentHelper.java:58)
        at com.awslabs.aws.greengrass.provisioner.interfaces.helpers.Operation.executeOrDisplayHelp(Operation.java:33)
        at com.awslabs.aws.greengrass.provisioner.interfaces.helpers.Operation.lambda$execute$1(Operation.java:21)
        at io.vavr.control.Try.onSuccess(Try.java:712)
        at com.awslabs.aws.greengrass.provisioner.interfaces.helpers.Operation.execute(Operation.java:21)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.lambda$null$2(AwsGreengrassProvisioner.java:91)
        at java.base/java.util.Optional.map(Optional.java:265)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.lambda$run$da6d192b$1(AwsGreengrassProvisioner.java:91)
        at io.vavr.control.Try.of(Try.java:75)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.run(AwsGreengrassProvisioner.java:86)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.runProvisioner(AwsGreengrassProvisioner.java:64)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.lambda$main$249538b4$1(AwsGreengrassProvisioner.java:32)
        at io.vavr.control.Try.of(Try.java:75)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.main(AwsGreengrassProvisioner.java:32)
[ERROR] AwsGreengrassProvisioner: Invalid resource configuration
```

Could you let me know how to attach one resource to several lambdas?

@timmattison After removing localVolumeResources from `function.defaults.conf`, I did as you said. However same error still occurs.

- function.conf in testFunction1 directory
```
conf {
  language = "PYTHON3_7"
  functionName = "testFunction1"
  localVolumeResources = [
    {
      name="test-storage"
      sourcePath = "/test"
      destinationPath = "/test" 
      readWrite = true
    }
  ]
 ...
}
```

- function.conf in testFunction2 directory
```
conf {
  language = "PYTHON3_7"
  functionName = "testFunction2"
  localVolumeResources = [
    {
      name="test-storage"
      sourcePath = "/test"
      destinationPath = "/test" 
      readWrite = true
    }
  ]
 ...
}
```

When I try to define local device(e.g. /dev/fb1) to each function.conf as link which you mentioned, completely same error occurs.

```
[INFO] BasicDeploymentHelper: Creating resource definition
[ERROR] BasicGreengrassHelper: Duplicate local device resource source paths defined ["/dev/fb1"].  Greengrass will not accept this configuration.
[ERROR] AwsGreengrassProvisioner: Invalid resource configuration
java.lang.RuntimeException: Invalid resource configuration
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicGreengrassHelper.disallowDuplicateLocalDeviceSourcePaths(BasicGreengrassHelper.java:923)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicGreengrassHelper.validateResourceDefinitionVersion(BasicGreengrassHelper.java:862)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicGreengrassHelper.createResourceDefinitionVersion(BasicGreengrassHelper.java:710)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicDeploymentHelper.execute(BasicDeploymentHelper.java:689)
        at com.awslabs.aws.greengrass.provisioner.implementations.helpers.BasicDeploymentHelper.execute(BasicDeploymentHelper.java:58)
        at com.awslabs.aws.greengrass.provisioner.interfaces.helpers.Operation.executeOrDisplayHelp(Operation.java:33)
        at com.awslabs.aws.greengrass.provisioner.interfaces.helpers.Operation.lambda$execute$1(Operation.java:21)
        at io.vavr.control.Try.onSuccess(Try.java:712)
        at com.awslabs.aws.greengrass.provisioner.interfaces.helpers.Operation.execute(Operation.java:21)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.lambda$null$2(AwsGreengrassProvisioner.java:91)
        at java.base/java.util.Optional.map(Optional.java:265)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.lambda$run$da6d192b$1(AwsGreengrassProvisioner.java:91)
        at io.vavr.control.Try.of(Try.java:75)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.run(AwsGreengrassProvisioner.java:86)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.runProvisioner(AwsGreengrassProvisioner.java:64)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.lambda$main$249538b4$1(AwsGreengrassProvisioner.java:32)
        at io.vavr.control.Try.of(Try.java:75)
        at com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner.main(AwsGreengrassProvisioner.java:32)
[ERROR] AwsGreengrassProvisioner: Invalid resource configuration
```

--- a/src/integration-test/java/GreengrassBuildWithDockerDeploymentsIT.java
+++ b/src/integration-test/java/GreengrassBuildWithDockerDeploymentsIT.java
@@ -4,7 +4,6 @@ import com.awslabs.aws.greengrass.provisioner.interfaces.helpers.IoHelper;
 import com.awslabs.aws.greengrass.provisioner.interfaces.helpers.ProcessHelper;
 import org.hamcrest.Matcher;
 import org.hamcrest.MatcherAssert;
-import org.jetbrains.annotations.NotNull;
 import org.junit.*;
 import org.junit.contrib.java.lang.system.ExpectedSystemExit;
 import org.slf4j.Logger;
@@ -25,6 +24,7 @@ import static org.hamcrest.CoreMatchers.*;
 
 // NOTE: Ignoring these tests since currently testcontainers throws a java.lang.OutOfMemoryError on every test
 public class GreengrassBuildWithDockerDeploymentsIT {
+    public static final String DOT_AWS = ".aws";
     private static final Matcher<Integer> EXIT_CODE_IS_ZERO = equalTo(0);
     private static final Matcher<Integer> EXIT_CODE_IS_NOT_ZERO = not(EXIT_CODE_IS_ZERO);
     private static Logger log = LoggerFactory.getLogger(GreengrassBuildWithDockerDeploymentsIT.class);
@@ -96,25 +96,25 @@ public class GreengrassBuildWithDockerDeploymentsIT {
     }
 
     private GenericContainer startAndGetContainer(String arguments) {
-        String hostAwsCredentialsPath = String.join("/", getHome(), ".aws");
+        String hostAwsCredentialsPath = String.join("/", getHome(), DOT_AWS);
         MountableFile hostAwsCredentialsMountableFile = MountableFile.forHostPath(new File(hostAwsCredentialsPath).toPath());
-        String containerAwsCredentialsPath = "/root/.aws";
+        String containerAwsCredentialsPath = String.join("", "/root/", DOT_AWS);
 
-        String hostFoundationPath = getHostPath("foundation");
+        String hostFoundationPath = GreengrassITShared.FOUNDATION;
         MountableFile hostFoundationMountableFile = MountableFile.forHostPath(new File(hostFoundationPath).toPath());
-        String containerFoundationPath = "/foundation";
+        String containerFoundationPath = String.join("", "/", GreengrassITShared.FOUNDATION);
 
-        String hostDeploymentsPath = getHostPath("deployments");
+        String hostDeploymentsPath = GreengrassITShared.DEPLOYMENTS;
         MountableFile hostDeploymentsMountableFile = MountableFile.forHostPath(new File(hostDeploymentsPath).toPath());
-        String containerDeploymentsPath = "/deployments";
+        String containerDeploymentsPath = String.join("", "/", GreengrassITShared.DEPLOYMENTS);
 
-        String hostFunctionsPath = getHostPath("functions");
+        String hostFunctionsPath = GreengrassITShared.FUNCTIONS;
         MountableFile hostFunctionsMountableFile = MountableFile.forHostPath(new File(hostFunctionsPath).toPath());
-        String containerFunctionsPath = "/functions";
+        String containerFunctionsPath = String.join("", "/", GreengrassITShared.FUNCTIONS);
 
-        String hostGgdsPath = getHostPath("ggds");
+        String hostGgdsPath = GreengrassITShared.GGDS;
         MountableFile hostGgdsMountableFile = MountableFile.forHostPath(new File(hostGgdsPath).toPath());
-        String containerGgdsPath = "/ggds";
+        String containerGgdsPath = String.join("", "/", GreengrassITShared.GGDS);
 
         GenericContainer genericContainer = new GenericContainer<>(getContainerName())
                 .withCopyFileToContainer(hostAwsCredentialsMountableFile, containerAwsCredentialsPath)
@@ -127,11 +127,6 @@ public class GreengrassBuildWithDockerDeploymentsIT {
         genericContainer.start();
 
         return genericContainer;
-    }
-
-    @NotNull
-    private String getHostPath(String foundation) {
-        return String.join("/", "..", "aws-greengrass-lambda-functions", foundation);
     }
 
     private void waitForContainerToFinish(GenericContainer genericContainer) {

--- a/src/integration-test/java/GreengrassBuildWithDockerDeploymentsIT.java
+++ b/src/integration-test/java/GreengrassBuildWithDockerDeploymentsIT.java
@@ -24,7 +24,6 @@ import java.util.concurrent.TimeUnit;
 import static org.hamcrest.CoreMatchers.*;
 
 // NOTE: Ignoring these tests since currently testcontainers throws a java.lang.OutOfMemoryError on every test
-@Ignore
 public class GreengrassBuildWithDockerDeploymentsIT {
     private static final Matcher<Integer> EXIT_CODE_IS_ZERO = equalTo(0);
     private static final Matcher<Integer> EXIT_CODE_IS_NOT_ZERO = not(EXIT_CODE_IS_ZERO);

--- a/src/integration-test/java/GreengrassBuildWithDockerDeploymentsIT.java
+++ b/src/integration-test/java/GreengrassBuildWithDockerDeploymentsIT.java
@@ -26,7 +26,6 @@ import static org.hamcrest.CoreMatchers.*;
 // NOTE: Ignoring these tests since currently testcontainers throws a java.lang.OutOfMemoryError on every test
 @Ignore
 public class GreengrassBuildWithDockerDeploymentsIT {
-    public static final String AWS_GREENGRASS_LAMBDA_FUNCTIONS_VIA_PARENT_DIRECTORY = "../aws-greengrass-lambda-functions";
     private static final Matcher<Integer> EXIT_CODE_IS_ZERO = equalTo(0);
     private static final Matcher<Integer> EXIT_CODE_IS_NOT_ZERO = not(EXIT_CODE_IS_ZERO);
     private static Logger log = LoggerFactory.getLogger(GreengrassBuildWithDockerDeploymentsIT.class);
@@ -94,7 +93,7 @@ public class GreengrassBuildWithDockerDeploymentsIT {
     }
 
     private String getContainerName() {
-        return String.join(":", GreengrassITShared.TIMMATTISON_AWS_GREENGRASS_PROVISIONER, getBranch());
+        return String.join(":", GreengrassITShared.DOCKERHUB_TIMMATTISON_AWS_GREENGRASS_PROVISIONER, getBranch());
     }
 
     private GenericContainer startAndGetContainer(String arguments) {
@@ -228,7 +227,7 @@ public class GreengrassBuildWithDockerDeploymentsIT {
     // Test set 11: Reuse Python 3 Hello World in another group with Docker
     @Test
     public void shouldReusePython3FunctionWithDocker() throws IOException {
-        File tempDeploymentConfFile = greengrassITShared.setupReusedFunctionDeploymentConf(Optional.of(AWS_GREENGRASS_LAMBDA_FUNCTIONS_VIA_PARENT_DIRECTORY), GreengrassITShared.HELLO_WORLD_PYTHON_3_PROD_PARTIAL, greengrassITShared.getGroupName());
+        File tempDeploymentConfFile = greengrassITShared.setupReusedFunctionDeploymentConf(Optional.empty(), GreengrassITShared.HELLO_WORLD_PYTHON_3_PROD_PARTIAL, greengrassITShared.getGroupName());
 
         // Build it
         runContainer(greengrassITShared.getPython3HelloWorldDeploymentCommand(Optional.empty()), EXIT_CODE_IS_ZERO);
@@ -240,7 +239,7 @@ public class GreengrassBuildWithDockerDeploymentsIT {
     // Test set 12: Reuse Java benchmark in another group with Docker
     @Test
     public void shouldReuseJavaFunctionWithDocker() throws IOException {
-        File tempDeploymentConfFile = greengrassITShared.setupReusedFunctionDeploymentConf(Optional.of(AWS_GREENGRASS_LAMBDA_FUNCTIONS_VIA_PARENT_DIRECTORY), GreengrassITShared.BENCHMARK_JAVA_PROD_PARTIAL, greengrassITShared.getGroupName());
+        File tempDeploymentConfFile = greengrassITShared.setupReusedFunctionDeploymentConf(Optional.empty(), GreengrassITShared.BENCHMARK_JAVA_PROD_PARTIAL, greengrassITShared.getGroupName());
 
         // Build it
         runContainer(greengrassITShared.getBenchmarkDeploymentCommand(Optional.empty()), EXIT_CODE_IS_ZERO);
@@ -252,7 +251,7 @@ public class GreengrassBuildWithDockerDeploymentsIT {
     // Test set 13: Fail to reuse Java benchmark in another group with Docker
     @Test
     public void shouldFailToReuseBroadPatternFunctionWithDocker() throws IOException {
-        File tempDeploymentConfFile = greengrassITShared.setupReusedFunctionDeploymentConf(Optional.of(AWS_GREENGRASS_LAMBDA_FUNCTIONS_VIA_PARENT_DIRECTORY), GreengrassITShared.BENCHMARK_PROD_PARTIAL, greengrassITShared.getGroupName());
+        File tempDeploymentConfFile = greengrassITShared.setupReusedFunctionDeploymentConf(Optional.empty(), GreengrassITShared.BENCHMARK_PROD_PARTIAL, greengrassITShared.getGroupName());
 
         // Build it
         runContainer(greengrassITShared.getBenchmarkDeploymentCommand(Optional.empty()), EXIT_CODE_IS_ZERO);

--- a/src/integration-test/java/GreengrassITShared.java
+++ b/src/integration-test/java/GreengrassITShared.java
@@ -66,7 +66,7 @@ class GreengrassITShared {
 
         FileUtils.copyDirectory(GreengrassITShared.MASTER_DEPLOYMENTS, GreengrassITShared.TEMP_DEPLOYMENTS);
         FileUtils.copyDirectory(GreengrassITShared.MASTER_FOUNDATION, GreengrassITShared.TEMP_FOUNDATION);
-       
+
         IOFileFilter ignoreVenvDirectories = FileFilterUtils.notFileFilter(FileFilterUtils.nameFileFilter("venv"));
 
         FileUtils.copyDirectory(GreengrassITShared.MASTER_FUNCTIONS, GreengrassITShared.TEMP_FUNCTIONS, ignoreVenvDirectories);

--- a/src/integration-test/java/GreengrassITShared.java
+++ b/src/integration-test/java/GreengrassITShared.java
@@ -5,6 +5,8 @@ import com.awslabs.aws.greengrass.provisioner.data.arguments.UpdateArguments;
 import com.awslabs.aws.greengrass.provisioner.interfaces.helpers.IoHelper;
 import com.awslabs.aws.greengrass.provisioner.interfaces.helpers.ThreadHelper;
 import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.filefilter.FileFilterUtils;
+import org.apache.commons.io.filefilter.IOFileFilter;
 import org.jetbrains.annotations.NotNull;
 
 import java.io.File;
@@ -15,13 +17,17 @@ class GreengrassITShared {
     static final String HELLO_WORLD_PYTHON_3_PROD_PARTIAL = "~HelloWorldPython3:PROD";
     static final String BENCHMARK_JAVA_PROD_PARTIAL = "~CDDBenchmarkJava:PROD";
     static final String BENCHMARK_PROD_PARTIAL = "~Benchmark~:PROD";
-    static final File MASTER_DEPLOYMENTS = new File("../aws-greengrass-lambda-functions/deployments");
-    static final File MASTER_FUNCTIONS = new File("../aws-greengrass-lambda-functions/functions");
-    static final File MASTER_FOUNDATION = new File("../aws-greengrass-lambda-functions/foundation");
-    static final String TIMMATTISON_AWS_GREENGRASS_PROVISIONER = "timmattison/aws-greengrass-provisioner";
-    static final File TEMP_DEPLOYMENTS = new File("deployments");
-    static final File TEMP_FUNCTIONS = new File("functions");
-    static final File TEMP_FOUNDATION = new File("foundation");
+    static final String LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT = String.join("/", "..", "aws-greengrass-lambda-functions");
+    static final String DEPLOYMENTS = "deployments";
+    static final String FUNCTIONS = "functions";
+    static final String FOUNDATION = "foundation";
+    static final File MASTER_DEPLOYMENTS = new File(String.join("/", LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT, DEPLOYMENTS));
+    static final File MASTER_FUNCTIONS = new File(String.join("/", LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT, FUNCTIONS));
+    static final File MASTER_FOUNDATION = new File(String.join("/", LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT, FOUNDATION));
+    static final String DOCKERHUB_TIMMATTISON_AWS_GREENGRASS_PROVISIONER = "timmattison/aws-greengrass-provisioner";
+    static final File TEMP_DEPLOYMENTS = new File(DEPLOYMENTS);
+    static final File TEMP_FUNCTIONS = new File(FUNCTIONS);
+    static final File TEMP_FOUNDATION = new File(FOUNDATION);
     static final File TEMP_CREDENTIALS = new File("credentials");
     /**
      * This is used to generate files that can be used in end-to-end tests
@@ -59,8 +65,11 @@ class GreengrassITShared {
         cleanDirectories();
 
         FileUtils.copyDirectory(GreengrassITShared.MASTER_DEPLOYMENTS, GreengrassITShared.TEMP_DEPLOYMENTS);
-        FileUtils.copyDirectory(GreengrassITShared.MASTER_FUNCTIONS, GreengrassITShared.TEMP_FUNCTIONS);
         FileUtils.copyDirectory(GreengrassITShared.MASTER_FOUNDATION, GreengrassITShared.TEMP_FOUNDATION);
+       
+        IOFileFilter ignoreVenvDirectories = FileFilterUtils.notFileFilter(FileFilterUtils.nameFileFilter("venv"));
+
+        FileUtils.copyDirectory(GreengrassITShared.MASTER_FUNCTIONS, GreengrassITShared.TEMP_FUNCTIONS, ignoreVenvDirectories);
     }
 
     String getGroupName() {
@@ -153,10 +162,10 @@ class GreengrassITShared {
 
     @NotNull
     private File writeTempDeploymentConfFile(Optional<String> optionalDirectoryPrefix, String tempDeploymentConfContents) throws IOException {
-        String directory = "deployments";
+        String directory = DEPLOYMENTS;
 
         if (optionalDirectoryPrefix.isPresent()) {
-            directory = optionalDirectoryPrefix.get() + "/" + directory;
+            directory = String.join("/", optionalDirectoryPrefix.get(), directory);
         }
 
         File tempDeploymentConfFile = File.createTempFile("temp-deployment-", ".conf", new File(directory));

--- a/src/integration-test/java/GreengrassITShared.java
+++ b/src/integration-test/java/GreengrassITShared.java
@@ -21,14 +21,17 @@ class GreengrassITShared {
     static final String DEPLOYMENTS = "deployments";
     static final String FUNCTIONS = "functions";
     static final String FOUNDATION = "foundation";
+    static final String GGDS = "ggds";
     static final File MASTER_DEPLOYMENTS = new File(String.join("/", LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT, DEPLOYMENTS));
     static final File MASTER_FUNCTIONS = new File(String.join("/", LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT, FUNCTIONS));
     static final File MASTER_FOUNDATION = new File(String.join("/", LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT, FOUNDATION));
+    static final File MASTER_GGDS = new File(String.join("/", LAMBDA_FUNCTIONS_DIRECTORY_VIA_PARENT, GGDS));
     static final String DOCKERHUB_TIMMATTISON_AWS_GREENGRASS_PROVISIONER = "timmattison/aws-greengrass-provisioner";
     static final File TEMP_DEPLOYMENTS = new File(DEPLOYMENTS);
     static final File TEMP_FUNCTIONS = new File(FUNCTIONS);
     static final File TEMP_FOUNDATION = new File(FOUNDATION);
     static final File TEMP_CREDENTIALS = new File("credentials");
+    static final File TEMP_GGDS = new File(GGDS);
     /**
      * This is used to generate files that can be used in end-to-end tests
      */
@@ -55,6 +58,7 @@ class GreengrassITShared {
         FileUtils.deleteDirectory(TEMP_FUNCTIONS);
         FileUtils.deleteDirectory(TEMP_FOUNDATION);
         FileUtils.deleteDirectory(TEMP_CREDENTIALS);
+        FileUtils.deleteDirectory(TEMP_GGDS);
     }
 
     static ThreadHelper getThreadHelper() {
@@ -70,6 +74,7 @@ class GreengrassITShared {
         IOFileFilter ignoreVenvDirectories = FileFilterUtils.notFileFilter(FileFilterUtils.nameFileFilter("venv"));
 
         FileUtils.copyDirectory(GreengrassITShared.MASTER_FUNCTIONS, GreengrassITShared.TEMP_FUNCTIONS, ignoreVenvDirectories);
+        FileUtils.copyDirectory(GreengrassITShared.MASTER_GGDS, GreengrassITShared.TEMP_GGDS, ignoreVenvDirectories);
     }
 
     String getGroupName() {

--- a/src/integration-test/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/DuplicateLocalResourceIT.java
+++ b/src/integration-test/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/DuplicateLocalResourceIT.java
@@ -1,0 +1,89 @@
+package com.awslabs.aws.greengrass.provisioner.implementations.helpers;
+
+import com.awslabs.aws.greengrass.provisioner.AwsGreengrassProvisioner;
+import com.awslabs.aws.greengrass.provisioner.data.resources.*;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DuplicateLocalResourceIT {
+    public static final String TEMPDIR_1 = "/tempdir1";
+    public static final String TMP_1 = "/tmp1";
+    public static final String TMP_2 = "/tmp2";
+    public static final String TMP = "/tmp";
+    public static final String TEMPDIR_2 = "/tempdir2";
+    private BasicGreengrassHelper basicGreengrassHelper;
+    private List<LocalDeviceResource> localDeviceResources = new ArrayList<>();
+    private List<LocalVolumeResource> localVolumeResources = new ArrayList<>();
+    private List<LocalS3Resource> localS3Resources = new ArrayList<>();
+    private List<LocalSageMakerResource> localSageMakerResources = new ArrayList<>();
+    private List<LocalSecretsManagerResource> localSecretsManagerResources = new ArrayList<>();
+
+    @Before
+    public void setup() {
+        basicGreengrassHelper = AwsGreengrassProvisioner.getInjector().getInstance(BasicGreengrassHelper.class);
+    }
+
+    @Test
+    public void shouldFilterOutDuplicateLocalVolumeResources() {
+        LocalVolumeResource localVolumeResource1 = ImmutableLocalVolumeResource.builder()
+                .sourcePath(TMP)
+                .destinationPath(TMP)
+                .isReadWrite(true)
+                .build();
+
+        LocalVolumeResource localVolumeResource2 = ImmutableLocalVolumeResource.copyOf(localVolumeResource1);
+
+        localVolumeResources.clear();
+        localVolumeResources.add(localVolumeResource1);
+        localVolumeResources.add(localVolumeResource2);
+
+        basicGreengrassHelper.createResourceDefinition(localDeviceResources, localVolumeResources, localS3Resources, localSageMakerResources, localSecretsManagerResources);
+    }
+
+    @Test
+    public void shouldFailWhenSourcePathsAreTheSameAndDestinationPathsAreDifferent() {
+        LocalVolumeResource localVolumeResource1 = ImmutableLocalVolumeResource.builder()
+                .sourcePath(TMP)
+                .destinationPath(TEMPDIR_1)
+                .isReadWrite(true)
+                .build();
+
+        LocalVolumeResource localVolumeResource2 = ImmutableLocalVolumeResource.builder()
+                .sourcePath(TMP)
+                .destinationPath(TEMPDIR_2)
+                .isReadWrite(true)
+                .build();
+
+        localVolumeResources.clear();
+        localVolumeResources.add(localVolumeResource1);
+        localVolumeResources.add(localVolumeResource2);
+
+        RuntimeException runtimeException = Assert.assertThrows(RuntimeException.class, () -> basicGreengrassHelper.createResourceDefinition(localDeviceResources, localVolumeResources, localS3Resources, localSageMakerResources, localSecretsManagerResources));
+        Assert.assertTrue(runtimeException.getMessage().contains("Invalid resource configuration"));
+    }
+
+    @Test
+    public void shouldNotFailWhenSourcePathsAreDifferentAndDestinationPathsAreTheSame() {
+        LocalVolumeResource localVolumeResource1 = ImmutableLocalVolumeResource.builder()
+                .sourcePath(TMP_1)
+                .destinationPath(TEMPDIR_1)
+                .isReadWrite(true)
+                .build();
+
+        LocalVolumeResource localVolumeResource2 = ImmutableLocalVolumeResource.builder()
+                .sourcePath(TMP_2)
+                .destinationPath(TEMPDIR_1)
+                .isReadWrite(true)
+                .build();
+
+        localVolumeResources.clear();
+        localVolumeResources.add(localVolumeResource1);
+        localVolumeResources.add(localVolumeResource2);
+
+        basicGreengrassHelper.createResourceDefinition(localDeviceResources, localVolumeResources, localS3Resources, localSageMakerResources, localSecretsManagerResources);
+    }
+}

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalDeviceResource.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalDeviceResource.java
@@ -3,10 +3,6 @@ package com.awslabs.aws.greengrass.provisioner.data.resources;
 import org.immutables.value.Value;
 
 @Value.Immutable
-public abstract class LocalDeviceResource {
-    public abstract String getName();
-
-    public abstract String getPath();
-
+public abstract class LocalDeviceResource implements LocalResource {
     public abstract boolean isReadWrite();
 }

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalResource.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalResource.java
@@ -1,0 +1,22 @@
+package com.awslabs.aws.greengrass.provisioner.data.resources;
+
+import java.util.Optional;
+
+public interface LocalResource {
+    default String getName() {
+        return getSafeName();
+    }
+
+    String getPath();
+
+    default String getSafeName() {
+        // Device names can't have special characters in them - https://docs.aws.amazon.com/greengrass/latest/apireference/createresourcedefinition-post.html
+        return Optional.of(getPath()
+                .replaceAll("[^a-zA-Z0-9:_-]", "-")
+                .replaceFirst("^-", "")
+                .replaceFirst("-$", "")
+                .trim())
+                .orElseThrow(() -> new RuntimeException("Path cannot be empty"));
+    }
+
+}

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalS3Resource.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalS3Resource.java
@@ -3,10 +3,6 @@ package com.awslabs.aws.greengrass.provisioner.data.resources;
 import org.immutables.value.Value;
 
 @Value.Immutable
-public abstract class LocalS3Resource {
-    public abstract String getName();
-
+public abstract class LocalS3Resource implements LocalResource {
     public abstract String getUri();
-
-    public abstract String getPath();
 }

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalSageMakerResource.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalSageMakerResource.java
@@ -3,10 +3,6 @@ package com.awslabs.aws.greengrass.provisioner.data.resources;
 import org.immutables.value.Value;
 
 @Value.Immutable
-public abstract class LocalSageMakerResource {
-    public abstract String getName();
-
+public abstract class LocalSageMakerResource implements LocalResource {
     public abstract String getArn();
-
-    public abstract String getPath();
 }

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalVolumeResource.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/data/resources/LocalVolumeResource.java
@@ -3,10 +3,12 @@ package com.awslabs.aws.greengrass.provisioner.data.resources;
 import org.immutables.value.Value;
 
 @Value.Immutable
-public abstract class LocalVolumeResource {
-    public abstract String getName();
-
+public abstract class LocalVolumeResource implements LocalResource {
     public abstract String getSourcePath();
+
+    public String getPath() {
+        return String.join("-", getSourcePath(), getDestinationPath());
+    }
 
     public abstract String getDestinationPath();
 

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicDeploymentHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicDeploymentHelper.java
@@ -698,7 +698,7 @@ public class BasicDeploymentHelper implements DeploymentHelper {
         ////////////////////////////
 
         log.info("Creating resource definition");
-        String resourceDefinitionVersionArn = greengrassHelper.createResourceDefinitionVersion(functionConfs);
+        String resourceDefinitionVersionArn = greengrassHelper.createResourceDefinitionFromFunctionConfs(functionConfs);
 
         /////////////////////////////////////////////////////////////////////////
         // Build the function definition for the Lambda function and a version //

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicFunctionHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicFunctionHelper.java
@@ -424,11 +424,7 @@ public class BasicFunctionHelper implements FunctionHelper {
 
             String path = temp.getString(PATH);
 
-            Optional<String> optionalName = getName(temp);
-            String name = makeNameSafe(path, optionalName);
-
             LocalDeviceResource localDeviceResource = ImmutableLocalDeviceResource.builder()
-                    .name(name)
                     .path(path)
                     .isReadWrite(temp.getBoolean(READ_WRITE))
                     .build();
@@ -453,11 +449,7 @@ public class BasicFunctionHelper implements FunctionHelper {
                     .recover(ConfigException.Missing.class, throwable -> sourcePath)
                     .get();
 
-            Optional<String> optionalName = getName(temp);
-            String name = makeNameSafe(sourcePath, optionalName);
-
             LocalVolumeResource localVolumeResource = ImmutableLocalVolumeResource.builder()
-                    .name(name)
                     .sourcePath(sourcePath)
                     .destinationPath(destinationPath)
                     .isReadWrite(temp.getBoolean(READ_WRITE))
@@ -479,11 +471,7 @@ public class BasicFunctionHelper implements FunctionHelper {
             String uri = temp.getString(URI);
             String path = temp.getString(PATH);
 
-            Optional<String> optionalName = getName(temp);
-            String name = makeNameSafe(path, optionalName);
-
             LocalS3Resource localS3Resource = ImmutableLocalS3Resource.builder()
-                    .name(name)
                     .uri(uri)
                     .path(path)
                     .build();
@@ -528,11 +516,7 @@ public class BasicFunctionHelper implements FunctionHelper {
                 throw new RuntimeException("SageMaker ARNs must be training job ARNs, not model ARNs or other types of ARNs [" + arn + "]");
             }
 
-            Optional<String> optionalName = getName(temp);
-            String name = makeNameSafe(path, optionalName);
-
             LocalSageMakerResource localSageMakerResource = ImmutableLocalSageMakerResource.builder()
-                    .name(name)
                     .arn(arn)
                     .path(path)
                     .build();
@@ -604,16 +588,6 @@ public class BasicFunctionHelper implements FunctionHelper {
         return Try.of(() -> Optional.of(config.getString("name")))
                 .recover(ConfigException.Missing.class, throwable -> Optional.empty())
                 .get();
-    }
-
-    private String makeNameSafe(String path, Optional<String> name) {
-        // Device names can't have special characters in them - https://docs.aws.amazon.com/greengrass/latest/apireference/createresourcedefinition-post.html
-        return Optional.of(name.orElse(path)
-                .replaceAll("[^a-zA-Z0-9:_-]", "-")
-                .replaceFirst("^-", "")
-                .replaceFirst("-$", "")
-                .trim())
-                .orElseThrow(() -> new RuntimeException("Name cannot be empty"));
     }
 
     private void setEnvironmentVariablesFromConf(ImmutableFunctionConf.Builder functionConfBuilder, Config config) {

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassHelper.java
@@ -16,6 +16,8 @@ import software.amazon.awssdk.services.iot.model.ResourceNotFoundException;
 
 import javax.inject.Inject;
 import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 public class BasicGreengrassHelper implements GreengrassHelper {
@@ -709,15 +711,35 @@ public class BasicGreengrassHelper implements GreengrassHelper {
                 .filter(FunctionConf::isGreengrassContainer)
                 .collect(Collectors.toList());
 
+        // Only get the resources for functions in the container
+        List<LocalDeviceResource> localDeviceResources = greengrassResourceHelper.flatMapResources(functionsInGreengrassContainer, FunctionConf::getLocalDeviceResources);
+        List<LocalVolumeResource> localVolumeResources = greengrassResourceHelper.flatMapResources(functionsInGreengrassContainer, FunctionConf::getLocalVolumeResources);
+        List<LocalS3Resource> localS3Resources = greengrassResourceHelper.flatMapResources(functionsInGreengrassContainer, FunctionConf::getLocalS3Resources);
+        List<LocalSageMakerResource> localSageMakerResources = greengrassResourceHelper.flatMapResources(functionsInGreengrassContainer, FunctionConf::getLocalSageMakerResources);
+        List<LocalSecretsManagerResource> localSecretsManagerResources = greengrassResourceHelper.flatMapResources(functionsInGreengrassContainer, FunctionConf::getLocalSecretsManagerResources);
+
+        return createResourceDefinition(localDeviceResources, localVolumeResources, localS3Resources, localSageMakerResources, localSecretsManagerResources);
+    }
+
+    public String createResourceDefinition(List<LocalDeviceResource> localDeviceResources, List<LocalVolumeResource> localVolumeResources, List<LocalS3Resource> localS3Resources, List<LocalSageMakerResource> localSageMakerResources, List<LocalSecretsManagerResource> localSecretsManagerResources) {
+        // Filter out duplicate devices (only the path needs to be the same)
+        localDeviceResources = localDeviceResources.stream()
+                .filter(distinctByKey(LocalDeviceResource::getPath))
+                .collect(Collectors.toList());
+
+        // Filter out duplicate volumes (only the source path needs to be the same)
+        localVolumeResources = localVolumeResources.stream()
+                .filter(distinctByKey(LocalVolumeResource::getPath))
+                .collect(Collectors.toList());
+
         List<Resource> resources = new ArrayList<>();
 
-        functionsInGreengrassContainer.forEach(functionConf -> {
-            resources.addAll(processLocalDeviceResources(functionConf));
-            resources.addAll(processLocalVolumeResources(functionConf));
-            resources.addAll(processLocalS3Resources(functionConf));
-            resources.addAll(processLocalSageMakerResources(functionConf));
-            resources.addAll(processLocalSecretsManagerResources(functionConf));
-        });
+        // Add those resources to the list
+        resources.addAll(processLocalDeviceResources(localDeviceResources));
+        resources.addAll(processLocalVolumeResources(localVolumeResources));
+        resources.addAll(processLocalS3Resources(localS3Resources));
+        resources.addAll(processLocalSageMakerResources(localSageMakerResources));
+        resources.addAll(processLocalSecretsManagerResources(localSecretsManagerResources));
 
         ResourceDefinitionVersion resourceDefinitionVersion = ResourceDefinitionVersion.builder()
                 .resources(resources)
@@ -733,6 +755,12 @@ public class BasicGreengrassHelper implements GreengrassHelper {
         CreateResourceDefinitionResponse createResourceDefinitionResponse = greengrassClient.createResourceDefinition(createResourceDefinitionRequest);
 
         return createResourceDefinitionResponse.latestVersionArn();
+    }
+
+    // From https://stackoverflow.com/a/27872852/796579
+    public <T> Predicate<T> distinctByKey(java.util.function.Function<? super T, ?> keyExtractor) {
+        Set<Object> seen = ConcurrentHashMap.newKeySet();
+        return t -> seen.add(keyExtractor.apply(t));
     }
 
     private boolean functionConfHasLocalResources(FunctionConf functionConf) {
@@ -760,8 +788,8 @@ public class BasicGreengrassHelper implements GreengrassHelper {
         return false;
     }
 
-    private List<Resource> processLocalSageMakerResources(FunctionConf functionConf) {
-        return functionConf.getLocalSageMakerResources().stream()
+    private List<Resource> processLocalSageMakerResources(List<LocalSageMakerResource> localSageMakerResources) {
+        return localSageMakerResources.stream()
                 .map(this::processLocalSageMakerResource)
                 .collect(Collectors.toList());
     }
@@ -779,8 +807,8 @@ public class BasicGreengrassHelper implements GreengrassHelper {
         return createResource(resourceDataContainer, localSageMakerResource.getName(), localSageMakerResource.getName());
     }
 
-    private List<Resource> processLocalSecretsManagerResources(FunctionConf functionConf) {
-        return functionConf.getLocalSecretsManagerResources().stream()
+    private List<Resource> processLocalSecretsManagerResources(List<LocalSecretsManagerResource> localSecretsManagerResources) {
+        return localSecretsManagerResources.stream()
                 .map(this::processLocalSecretsManagerResource)
                 .collect(Collectors.toList());
     }
@@ -797,8 +825,8 @@ public class BasicGreengrassHelper implements GreengrassHelper {
         return createResource(resourceDataContainer, localSecretsManagerResource.getResourceName(), localSecretsManagerResource.getResourceName());
     }
 
-    private List<Resource> processLocalS3Resources(FunctionConf functionConf) {
-        return functionConf.getLocalS3Resources().stream()
+    private List<Resource> processLocalS3Resources(List<LocalS3Resource> localS3Resources) {
+        return localS3Resources.stream()
                 .map(this::processLocalS3Resource)
                 .collect(Collectors.toList());
     }
@@ -816,8 +844,8 @@ public class BasicGreengrassHelper implements GreengrassHelper {
         return createResource(resourceDataContainer, localS3Resource.getName(), localS3Resource.getName());
     }
 
-    private List<Resource> processLocalVolumeResources(FunctionConf functionConf) {
-        return functionConf.getLocalVolumeResources().stream()
+    private List<Resource> processLocalVolumeResources(List<LocalVolumeResource> localVolumeResources) {
+        return localVolumeResources.stream()
                 .map(this::processLocalVolumeResource)
                 .collect(Collectors.toList());
     }
@@ -840,8 +868,8 @@ public class BasicGreengrassHelper implements GreengrassHelper {
         return createResource(resourceDataContainer, localVolumeResource.getName(), localVolumeResource.getName());
     }
 
-    private List<Resource> processLocalDeviceResources(FunctionConf functionConf) {
-        return functionConf.getLocalDeviceResources().stream()
+    private List<Resource> processLocalDeviceResources(List<LocalDeviceResource> localDeviceResources) {
+        return localDeviceResources.stream()
                 .map(this::processLocalDeviceResource)
                 .collect(Collectors.toList());
     }

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassHelper.java
@@ -697,7 +697,7 @@ public class BasicGreengrassHelper implements GreengrassHelper {
     }
 
     @Override
-    public String createResourceDefinitionVersion(List<FunctionConf> functionConfs) {
+    public String createResourceDefinitionFromFunctionConfs(List<FunctionConf> functionConfs) {
         // Log that the local resources from functions outside of the Greengrass container will be scrubbed
         functionConfs.stream()
                 .filter(functionConf -> !functionConf.isGreengrassContainer())

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassResourceHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassResourceHelper.java
@@ -1,5 +1,6 @@
 package com.awslabs.aws.greengrass.provisioner.implementations.helpers;
 
+import com.awslabs.aws.greengrass.provisioner.data.conf.FunctionConf;
 import com.awslabs.aws.greengrass.provisioner.interfaces.helpers.GreengrassResourceHelper;
 import io.vavr.control.Try;
 import org.slf4j.Logger;
@@ -58,6 +59,14 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
                 .collect(Collectors.toList());
 
         disallowDuplicateSecretsManagerSecrets(localSecretsManagerResources);
+    }
+
+    @Override
+    public <R> List<R> flatMapResources(List<FunctionConf> functionConfs, java.util.function.Function<FunctionConf, List<R>> method) {
+        return functionConfs.stream()
+                .map(method)
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
     }
 
     private void disallowDuplicateSecretsManagerSecrets(List<SecretsManagerSecretResourceData> resources) {
@@ -157,5 +166,17 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
         }
 
         return Optional.empty();
+    }
+
+    private Optional<List<Map.Entry<String, Set<String>>>> findMultipleDestinations(Map<String, Set<String>> inputMap) {
+        List<Map.Entry<String, Set<String>>> duplicates = inputMap.entrySet().stream()
+                .filter(entry -> entry.getValue().size() > 1)
+                .collect(Collectors.toList());
+
+        if (duplicates.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(duplicates);
     }
 }

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassResourceHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/implementations/helpers/BasicGreengrassResourceHelper.java
@@ -2,17 +2,13 @@ package com.awslabs.aws.greengrass.provisioner.implementations.helpers;
 
 import com.awslabs.aws.greengrass.provisioner.data.conf.FunctionConf;
 import com.awslabs.aws.greengrass.provisioner.interfaces.helpers.GreengrassResourceHelper;
-import io.vavr.control.Try;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import software.amazon.awssdk.services.greengrass.model.*;
 
 import javax.inject.Inject;
-import java.io.*;
 import java.util.*;
 import java.util.stream.Collectors;
-
-import static com.google.common.io.Resources.getResource;
 
 public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
     private final Logger log = LoggerFactory.getLogger(BasicGreengrassHelper.class);
@@ -23,37 +19,38 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
 
     @Override
     public void validateResourceDefinitionVersion(ResourceDefinitionVersion resourceDefinitionVersion) {
-        List<Resource> resourceList = resourceDefinitionVersion.resources();
+        List<Resource> resources = resourceDefinitionVersion.resources();
 
-        List<LocalDeviceResourceData> localDeviceResources = resourceList.stream()
+        List<LocalDeviceResourceData> localDeviceResources = resources.stream()
                 .map(res -> res.resourceDataContainer().localDeviceResourceData())
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
-        disallowDuplicateLocalDeviceSourcePaths(localDeviceResources);
+        disallowDuplicateDeviceSourcePaths(localDeviceResources);
+        disallowDeviceSourcePathsNotInDev(localDeviceResources);
 
-        List<LocalVolumeResourceData> localVolumeResources = resourceList.stream()
+        List<LocalVolumeResourceData> localVolumeResources = resources.stream()
                 .map(res -> res.resourceDataContainer().localVolumeResourceData())
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
-        disallowDuplicateLocalVolumeSourcePaths(localVolumeResources);
+        disallowDuplicateVolumeSourcePathsWithDifferentDestinationPaths(localVolumeResources);
 
-        List<S3MachineLearningModelResourceData> localS3Resources = resourceList.stream()
+        List<S3MachineLearningModelResourceData> localS3Resources = resources.stream()
                 .map(res -> res.resourceDataContainer().s3MachineLearningModelResourceData())
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
         disallowDuplicateS3DestinationPaths(localS3Resources);
 
-        List<SageMakerMachineLearningModelResourceData> localSageMakerResources = resourceList.stream()
+        List<SageMakerMachineLearningModelResourceData> localSageMakerResources = resources.stream()
                 .map(res -> res.resourceDataContainer().sageMakerMachineLearningModelResourceData())
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
         disallowDuplicateSageMakerDestinationPaths(localSageMakerResources);
 
-        List<SecretsManagerSecretResourceData> localSecretsManagerResources = resourceList.stream()
+        List<SecretsManagerSecretResourceData> localSecretsManagerResources = resources.stream()
                 .map(res -> res.resourceDataContainer().secretsManagerSecretResourceData())
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
@@ -74,7 +71,7 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
                 .map(SecretsManagerSecretResourceData::arn)
                 .collect(Collectors.toList());
 
-        Optional<List<String>> duplicates = findDuplicates(secretArns);
+        Optional<List<String>> duplicates = findMultipleDestinations(secretArns);
 
         if (duplicates.isPresent()) {
             String duplicatesString = duplicates.get().stream()
@@ -86,36 +83,59 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
         }
     }
 
-    private void disallowDuplicateLocalDeviceSourcePaths(List<LocalDeviceResourceData> resources) {
+    private void disallowDuplicateDeviceSourcePaths(List<LocalDeviceResourceData> resources) {
         List<String> sourcePaths = resources.stream()
                 .map(LocalDeviceResourceData::sourcePath)
                 .collect(Collectors.toList());
 
-        Optional<List<String>> duplicates = findDuplicates(sourcePaths);
+        Optional<List<String>> duplicates = findMultipleDestinations(sourcePaths);
 
-        if (duplicates.isPresent()) {
-            String duplicatesString = duplicates.get().stream()
-                    .map(string -> "\"" + string + "\"")
-                    .collect(Collectors.joining(", "));
-
-            log.error("Duplicate local device resource source paths defined [" + duplicatesString + "].  Greengrass will not accept this configuration.");
-            throw new RuntimeException("Invalid resource configuration");
+        if (!duplicates.isPresent()) {
+            return;
         }
+
+        String duplicatesString = duplicates.get().stream()
+                .map(string -> "\"" + string + "\"")
+                .collect(Collectors.joining(", "));
+
+        log.error("Duplicate local device resource source paths defined [" + duplicatesString + "].  Greengrass will not accept this configuration.");
+        throw new RuntimeException("Invalid resource configuration");
     }
 
-    private void disallowDuplicateLocalVolumeSourcePaths(List<LocalVolumeResourceData> resources) {
-        List<String> sourcePaths = resources.stream()
-                .map(LocalVolumeResourceData::sourcePath)
+    private void disallowDeviceSourcePathsNotInDev(List<LocalDeviceResourceData> resources) {
+        List<String> invalidSourcePaths = resources.stream()
+                .map(LocalDeviceResourceData::sourcePath)
+                .filter(sourcePath -> !sourcePath.startsWith("/dev"))
                 .collect(Collectors.toList());
 
-        Optional<List<String>> duplicates = findDuplicates(sourcePaths);
+        if (invalidSourcePaths.size() == 0) {
+            return;
+        }
 
-        if (duplicates.isPresent()) {
-            String duplicatesString = duplicates.get().stream()
-                    .map(string -> "\"" + string + "\"")
+        String invalidString = invalidSourcePaths.stream()
+                .map(string -> "\"" + string + "\"")
+                .collect(Collectors.joining(", "));
+
+        log.error("Local device resource source paths were specified outside of /dev [" + invalidString + "].  Greengrass will not accept this configuration.");
+        throw new RuntimeException("Invalid resource configuration");
+    }
+
+    private void disallowDuplicateVolumeSourcePathsWithDifferentDestinationPaths(List<LocalVolumeResourceData> resources) {
+        Map<String, Set<String>> map = new HashMap<>();
+
+        // Find the destination paths for each source path
+        resources.forEach(resource -> map.computeIfAbsent(resource.sourcePath(), x -> new HashSet<>()).add(resource.destinationPath()));
+
+        // Isolate the source paths with more than one destination path
+        Optional<List<Map.Entry<String, Set<String>>>> multipleDestinations = findMultipleDestinations(map);
+
+        if (multipleDestinations.isPresent()) {
+            String multipleDestinationsString = multipleDestinations.get().stream()
+                    .map(entry -> "Source [" + entry.getKey() + "], destinations [" + String.join(", ", entry.getValue()) + "]")
                     .collect(Collectors.joining(", "));
 
-            log.error("Duplicate local volume resource source paths defined [" + duplicatesString + "].  Greengrass will not accept this configuration.");
+            log.error("Duplicate local volume resource destination paths defined [" + multipleDestinationsString + "].  Greengrass will not accept this configuration.");
+
             throw new RuntimeException("Invalid resource configuration");
         }
     }
@@ -125,7 +145,7 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
                 .map(SageMakerMachineLearningModelResourceData::destinationPath)
                 .collect(Collectors.toList());
 
-        Optional<List<String>> duplicates = findDuplicates(sourcePaths);
+        Optional<List<String>> duplicates = findMultipleDestinations(sourcePaths);
 
         if (duplicates.isPresent()) {
             String duplicatesString = duplicates.get().stream()
@@ -142,7 +162,7 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
                 .map(S3MachineLearningModelResourceData::destinationPath)
                 .collect(Collectors.toList());
 
-        Optional<List<String>> duplicates = findDuplicates(sourcePaths);
+        Optional<List<String>> duplicates = findMultipleDestinations(sourcePaths);
 
         if (duplicates.isPresent()) {
             String duplicatesString = duplicates.get().stream()
@@ -154,7 +174,7 @@ public class BasicGreengrassResourceHelper implements GreengrassResourceHelper {
         }
     }
 
-    private Optional<List<String>> findDuplicates(List<String> inputList) {
+    private Optional<List<String>> findMultipleDestinations(List<String> inputList) {
         List<String> outputList = new ArrayList<>(inputList);
         Set<String> deduplicatedList = new HashSet<>(inputList);
 

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/interfaces/helpers/GreengrassHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/interfaces/helpers/GreengrassHelper.java
@@ -60,7 +60,7 @@ public interface GreengrassHelper {
 
     DeploymentStatus waitForDeploymentStatusToChange(String groupId, String deploymentId);
 
-    String createResourceDefinitionVersion(List<FunctionConf> functionConfs);
+    String createResourceDefinitionFromFunctionConfs(List<FunctionConf> functionConfs);
 
     Device getDevice(String thingName);
 

--- a/src/main/java/com/awslabs/aws/greengrass/provisioner/interfaces/helpers/GreengrassResourceHelper.java
+++ b/src/main/java/com/awslabs/aws/greengrass/provisioner/interfaces/helpers/GreengrassResourceHelper.java
@@ -1,7 +1,12 @@
 package com.awslabs.aws.greengrass.provisioner.interfaces.helpers;
 
+import com.awslabs.aws.greengrass.provisioner.data.conf.FunctionConf;
 import software.amazon.awssdk.services.greengrass.model.ResourceDefinitionVersion;
+
+import java.util.List;
 
 public interface GreengrassResourceHelper {
     void validateResourceDefinitionVersion(ResourceDefinitionVersion resourceDefinitionVersion);
+
+    <R> List<R> flatMapResources(List<FunctionConf> functionConfs, java.util.function.Function<FunctionConf, List<R>> method);
 }

